### PR TITLE
First plugin implementation

### PR DESCRIPTION
**Pull request by MetCoder95**
PR includes the first implementation of the plugin. This plugin comes as a solution to fastify's issue [#2619](https://github.com/fastify/fastify/issues/2619).

More PRs will come with future features like CI-CD, issue template, PR template, and so on.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,69 @@
 # fastify-get-head
-Fastify plugin to set a HEAD route handler for each GET handler previously registered
+
+[![Greenkeeper badge](https://badges.greenkeeper.io/fastify/fastify-cookie.svg)](https://greenkeeper.io/)
```
Comment (mcollina) [inline README.md:3]: remove greenkeeper badge, it's out of date and greenkeeper is not a thing anymore

File: README.md [hunk 2]
```diff
@@ -1,2 +1,69 @@
...
+
+<!-- [![Build Status](https://travis-ci.org/fastify/fastify-cookie.svg?branch=master)](https://travis-ci.org/fastify/fastify-cookie) -->
```
Comment (mcollina) [inline README.md:5]: Update this to github actions

File: README.md [hunk 3]
```diff
@@ -1,2 +1,69 @@
...
+
+A plugin for [Fastify](http://fastify.io/) that adds support for setting a `HEAD` route for each `GET` one previously registered.
+
+This plugin's cookie parsing works via Fastify's `onRoute` hook. When a new route is registered, the plugin will try to set a new `HEAD` route if the registered route is for a `GET` method.
+
+`fastify-cookie` v3 only supports Fastify@3.x.
```
Comment (mcollina) [inline README.md:11]: This is out of date

File: README.md [hunk 4]
```diff
@@ -1,2 +1,69 @@
...
+
+<!-- Setted once https://github.com/fastify/fastify/commit/449fc5c25e7d2e277e68bb3f4416cfd858a71cec is released -->
+
+## Example
+
+```js
+const fastify = require('fastify')();
+
+fastify.register(require('fastify-get-head'), {
+  ignorePaths: ['/api/ignore', /\/api\/ignore\/too/], // For ignoring specific paths
```
Comment (mcollina) [inline README.md:21]: Does this support regexps too?

File: README.md [hunk 5]
```diff
@@ -1,2 +1,69 @@
...
+// Same as doing
+
+fastify.head('/', (req, reply) => {
+  reply.headers({
+    ['content-length']: Buffer.from(JSON.stringify({ hello: 'world' })).byteLength
+    ['content-type']: 'application/json'
+  });
+
+  reply.status(200).send(null);
+});
```
Comment (mcollina) [inline README.md:38]: Maybe keep this commented out.

File: index.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
+const fp = require('fastify-plugin')
```
Comment (mcollina) [inline index.js:1]: add use strict at top.

File: index.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    // Only GET routes and allowed paths
+    if (isNotGetMethod || shouldIgnore) return
+
+    fastify.route({
+      ...routeConfig,
+      path,
+      handler,
+      method: 'HEAD',
+      onSend: onSendHandler
```
Comment (mcollina) [inline index.js:59]: You might have an `onSend` in `routeConfig`. If it is present, it should be converted to an array and add as the latest value.

File: index.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    })
+  }
+}
+
+function plugin (fastify, opts = {}, done) {
```
Comment (mcollina) [inline index.js:64]: remove the default value, it is not needed

File: package.json [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    "head"
+  ],
+  "author": "Carlos Fuentes <carlos.fuentes.rios.95@gmail.com>",
+  "license": "MIT",
+  "devDependencies": {
+    "@types/node": "^14.10.3",
+    "fastify": "^3.7.0",
+    "pre-commit": "^1.2.2",
+    "snazzy": "^8.0.0",
+    "standard": "^15.0.0",
```
Comment (mcollina) [inline package.json:28]: update this

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by mcollina** [comment]
reviewed!

File: README.md (hunk 4 above)
Comment (MetCoder95) [reply README.md:21]: Yes it does, I added support for regexps and string, so it's possible to use either or both of them. See [here](https://github.com/MetCoder95/fastify-get-head/pull/1/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R3)

File: index.js (hunk 7 above)
Comment (MetCoder95) [reply index.js:59]: True I missed that, thanks!

**Comment by MetCoder95** [comment]
> reviewed!

Thanks a lot, I'll address the requested changes over the week 😄

File: README.md (hunk 3 above)
Comment (MetCoder95) [reply README.md:11]: Sure, just to double-check, I think the plugin is coupled to the version where this [commit](https://github.com/fastify/fastify/commit/449fc5c25e7d2e277e68bb3f4416cfd858a71cec) is released, isn't it?

File: package.json (hunk 9 above)
Comment (MetCoder95) [reply package.json:28]: Done, please see [bb78a12](https://github.com/MetCoder95/fastify-get-head/pull/1/commits/bb78a120746ff60418d25a4e80cb7436e160ac33)

File: index.js (hunk 8 above)
Comment (MetCoder95) [reply index.js]: Done, please see [bb78a12](https://github.com/MetCoder95/fastify-get-head/pull/1/commits/bb78a120746ff60418d25a4e80cb7436e160ac33)

File: index.js (hunk 7 above)
Comment (MetCoder95) [reply index.js]: Done, please see [8c2b06f](https://github.com/MetCoder95/fastify-get-head/pull/1/commits/8c2b06f0a081d4ab75c70a1db999f24b6a654797)

**Comment by MetCoder95** [comment]
I addressed all the changes, the tests are failing as the most recent version of fastify doesn't include the support for the `HEAD` method, [this](https://github.com/fastify/fastify/commit/449fc5c25e7d2e277e68bb3f4416cfd858a71cec) in specific. Once released I'll update with the new versions 😄

**Comment by mcollina** [comment]
I've just shipped fastify v3.8.0 with all the latest changes, update :)

**Comment by MetCoder95** [comment]
Thanks a lot! Done! Please have a final look! :)
If everything ok, I'll publish the plugin by tomorrow! 🙏

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**